`test_accept_invalid_certificate/test.py::test_strict_reject_with_config` has failed

Report: https://s3.amazonaws.com/clickhouse-test-reports/67560/5d0c179ad190dcedf85a32821ebf6d25ec2b53a6/integration_tests__asan__old_analyzer__[4_6].html

https://s3.amazonaws.com/clickhouse-test-reports/0/992e70337f3b3f7e64388142cb19beba3e4b2312/integration_tests__asan__[2_4].html

```
=================================== FAILURES ===================================
________________________ test_strict_reject_with_config ________________________
[gw3] linux -- Python 3.10.12 /usr/bin/python3

    def test_strict_reject_with_config():
        with pytest.raises(Exception) as err:
            execute_query_native(node1, "SELECT 1", config_accept)
>       assert "alert certificate required" in str(err.value)
E       AssertionError: assert 'alert certificate required' in 'Client failed! Return code: 210, stderr: Code: 210. DB::NetException: Net Exception: Socket is not connected (172.16.3.3:9440). (NETWORK_ERROR)\n\n'
E        +  where 'Client failed! Return code: 210, stderr: Code: 210. DB::NetException: Net Exception: Socket is not connected (172.16.3.3:9440). (NETWORK_ERROR)\n\n' = str(QueryRuntimeException('Client failed! Return code: 210, stderr: Code: 210. DB::NetException: Net Exception: Socket is not connected (172.16.3.3:9440). (NETWORK_ERROR)\n\n'))
E        +    where QueryRuntimeException('Client failed! Return code: 210, stderr: Code: 210. DB::NetException: Net Exception: Socket is not connected (172.16.3.3:9440). (NETWORK_ERROR)\n\n') = <ExceptionInfo QueryRuntimeException('Client failed! Return code: 210, stderr: Code: 210. DB::NetException: Net Exception: Socket is not connected (172.16.3.3:9440). (NETWORK_ERROR)\n\n') tblen=5>.value

```

--- a/base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp
+++ b/base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp
@@ -311,7 +311,7 @@ int SecureSocketImpl::sendBytes(const void* buffer, int length, int flags)
 	while (mustRetry(rc, remaining_time));
 	if (rc <= 0)
 	{
-		// At this stage we still can have last not yet recieved SSL message containing SSL error
+		// At this stage we still can have last not yet received SSL message containing SSL error
 		// so make a read to force SSL to process possible SSL error
 		if (SSL_get_error(_pSSL, rc) == SSL_ERROR_SYSCALL && SocketImpl::lastError() == POCO_ECONNRESET)
 		{

--- a/base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp
+++ b/base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp
@@ -311,6 +311,14 @@ int SecureSocketImpl::sendBytes(const void* buffer, int length, int flags)
 	while (mustRetry(rc, remaining_time));
 	if (rc <= 0)
 	{
+		// At this stage we still can have last not yet recieved SSL message containing SSL error
+		// so make a read to force SSL to process possible SSL error
+		if (SSL_get_error(_pSSL, rc) == SSL_ERROR_SYSCALL && SocketImpl::lastError() == POCO_ECONNRESET)
+		{
+			char c = 0;
+			SSL_read(_pSSL, &c, 1);
+		}
+
 		rc = handleError(rc);
 		if (rc == 0) throw SSLConnectionUnexpectedlyClosedException();
 	}

--- a/src/Client/Connection.cpp
+++ b/src/Client/Connection.cpp
@@ -145,6 +145,9 @@ void Connection::connect(const ConnectionTimeouts & timeouts)
                 /// work we need to pass host name separately. It will be send into TLS Hello packet to let
                 /// the server know which host we want to talk with (single IP can process requests for multiple hosts using SNI).
                 static_cast<Poco::Net::SecureStreamSocket*>(socket.get())->setPeerHostName(host);
+                /// we want to postpone SSL handshake until first read or write operation
+                /// so any errors during negotiation would be properly processed
+                static_cast<Poco::Net::SecureStreamSocket*>(socket.get())->setLazyHandshake(true);
 #else
                 throw Exception(ErrorCodes::SUPPORT_IS_DISABLED, "tcp_secure protocol is disabled because poco library was built without NetSSL support.");
 #endif

